opengovsg/FormSG
# refactor: convert MailService to a class based Typescript implementation

## Problem

This PR converts the existing `MailService` to a class based Typescript implementation. 
A class based implementation is chosen to increase the testability of the service itself. For example, the mail transporter can now be easily mocked in tests.

In addition, `MailService` now takes on the responsibility of generating the mails' HTML instead of the controllers using `ejs`.

The new MailService now has the following exposed interfaces:
- `MailService#sendVerificationOtp`
- `MailService#sendLoginOtp`
- `MailService#sendSubmissionToAdmin`
- `MailService#sendAutoReplyEmails`

Tests are also finally added for `MailService`.

Related to #22

## Solution

**Features**:

- `MailService` is now a class based implementation
- Retries for mail is now removed, as we do not need retries anymore, since SES can do it for us. If we need to add retries in the future, we can also look into using [`promise-retry`](https://www.npmjs.com/package/promise-retry) or other libraries instead of hand rolling one ourselves.
- `utils/autoreply-pdf.js` is now deleted since `utils/mail` has taken over that responsibility.

**Fixes that snuck in**:
- Fix un-awaited `expectAsync` in one of `SmsService`'s tests

## Tests
- [ ] Create an email form, and add an email field with email verification and confirmation.
  1. Enter your email into the email form and verify your email. The email should be sent to the given email and the OTP should validate successfully.
  2. Submit the form. You (as form admin) should receive the usual form submission email with the correct answers. You (as the form filler) should also receive an acknowledgement email. 

# Changes from the PR that got merged into this PR:
## Problem

Whilst migrating `utils/constants.js`, realize that all it was holding were mail constants, so I sought to extract them into a new `constants/mail` directory. In doing so, realized an opportunistic refactor to add more responsibility to `MailService` and less on the controllers themselves.

In this PR, `MailService` hides the base `MailService#sendNodeMail` function and adds more specific functions for the various controllers to use:

- `MailService#sendVerificationOtp` used by `verification.service`
- `MailService#sendLoginOtp` used by `authentication.server.controller` (should be extracted to a new `AuthService` when that step happens)
- `MailService#sendSubmissionToAdmin` used by `email-submissions.server.controller` (should be extracted to a new `SubmissionService` when that step happens)
- `MailService#sendAutoReplyEmail` used by `submissions.server.controller` (same as above)

Also added comprehensive tests to `MailService`. 

Related to #8

## Solution

**Features**:

- Add more functions to MailServer and use them in the respective controllers/services
- Extract `utils/constants.js` into new `constants` directory containing `constants/mail` and `constants/filesize`

**Fixes**:
- Fix email guard if recipient mail is array (would throw error if recipient mail is passed as an array last time)
